libcontainer/dmz/dmz_linux.go

```diff
@@ -6,20 +6,32 @@ package dmz
 import (
 	"bytes"
 	"debug/elf"
-	_ "embed"
+	"embed"
 	"os"
+	"sync"
 
 	"github.com/sirupsen/logrus"
 )
 
-// Try to build the runc-dmz binary. If it fails, replace it with an empty file
-// (this will trigger us to fall back to a clone of /proc/self/exe). Yeah, this
-// is a bit ugly but it makes sure that weird cross-compilation setups don't
-// break because of runc-dmz.
+// Try to build the runc-dmz binary on "go generate". If it fails (or
+// libcontainer is being imported as a library), the embed.FS will not contain
+// the file, which will then cause us to fall back to a clone of
+// /proc/self/exe.
 //
-//go:generate sh -c "make -B runc-dmz || echo -n >runc-dmz"
-//go:embed runc-dmz
-var runcDmzBinary []byte
+// There is an empty file called runc-dmz-dummy-file.txt in libcontainer/dmz in
+// order to work around the restriction that go:embed requires at least one
+// file to match the pattern.
```

Comment:
A nit. AFAIK we don't need `runc-dmz-dummy-file.txt` in `libcontainer/dmz` anymore since we have `.gitignore` in there (which, by the way, is also being embedded).

Or, we need to do
```diff
-//go:embed binary
+//go:embed binary/runc-*
```
so that `.gitignore` is not included. 

Either way is fine.

Reply:
Actually `.gitignore` is not embedded [because files starting with `_` and `.` are not included by `go:embed` when embedding a directory unless you use the `all:` prefix](https://pkg.go.dev/embed#hdr-Directives):

> If a pattern names a directory, all files in the subtree rooted at that directory are embedded (recursively), except that files with names beginning with ‘.’ or ‘_’ are excluded. [...] If a pattern begins with the prefix ‘all:’, then the rule for walking directories is changed to include those files beginning with ‘.’ or ‘_’. For example, ‘all:image’ embeds both ‘image/.tempfile’ and ‘image/dir/.tempfile’.

If you remove `runc-dmz` and `dummy-file.txt` as well as the `go:generate` line in `libcontainer/dmz/dmz_linux.go` (to emulate using `go get` on runc), you'll find that the build fails with the same error as before.